libz-rs-sys/src/gz.rs

```diff
@@ -535,6 +535,67 @@ pub unsafe extern "C-unwind" fn gzerror(file: gzFile, errnum: *mut c_int) -> *co
     }
 }
 
+/// Clear the error and end-of-file state for `file`.
+///
+/// # Arguments
+///
+/// * `file` - A gzip file handle, or null
+///
+/// # Safety
+///
+/// `file`, if non-null, must be an open file handle obtained from [`gzopen`] or [`gzdopen`].
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzclearerr))]
+pub unsafe extern "C-unwind" fn gzclearerr(file: gzFile) {
+    // Get internal structure and check integrity
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return;
+    };
+    if state.mode != GzMode::GZ_READ && state.mode != GzMode::GZ_WRITE {
+        return;
+    }
+
+    // Clear error and EOF
+    if state.mode == GzMode::GZ_READ {
+        state.eof = false;
+        state.past = false;
+    }
+
+    // Safety: we've checked state, and gz_error supports a null message argument.
+    unsafe { gz_error(state, Z_OK, ptr::null()) };
+}
+
+/// Check whether a read operation has tried to read beyond the end of `file`.
+///
+/// # Returns
+///
+/// * 1 if the end-of-file indicator is set. Note that this indicator is set only
+///   if a read tries to go past the end of the input. If the last read request
+///   attempted to read exactly the number of bytes remaining in the file, the
+///   end-of-file indicator will not be set.
+/// * 0 the end-of-file indicator is not set or `file` is null
+///
+/// # Arguments
+///
+/// * `file` - A gzip file handle, or null
+///
+/// # Safety
+///
+/// `file`, if non-null, must be an open file handle obtained from [`gzopen`] or [`gzdopen`].
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzeof))]
+pub unsafe extern "C-unwind" fn gzeof(file: gzFile) -> c_int {
+    // Get internal structure and check integrity
+    let Some(state) = (unsafe { file.cast::<GzState>().as_ref() }) else {
+        return 0;
+    };
+    if state.mode != GzMode::GZ_READ {
```

Comment:
in zlib-ng, this condition is `    if (state->mode != GZ_READ && state->mode != GZ_WRITE)`

in that case you could even write something like this

```rust
match state.mode {
    GzMode::GZ_READ => state.past as _,
    GzMode::GZ_WRITE | GzMode::GZ_NONE | GzMode::GZ_APPEND => 0,
}
```

Reply:
That check in zlib-ng is redundant. It doesn't need to include the `&& state->mode != GZ_WRITE` because the next line is this:
```
    /* return end-of-file state */
    return state->mode == GZ_READ ? state->past : 0;
```

Reply:
ah right. let's add a test for that then and we should be good